std/jwt/mod.ts

```diff
@@ -0,0 +1,194 @@
+  verify as verifySignature,
+} from "./signature.ts";
+import { verify as verifyAlgorithm } from "./algorithm.ts";
+/*
+ * JWT §4.1: The following Claim Names are registered in the IANA
+ * "JSON Web Token Claims" registry established by Section 10.1. None of the
+ * claims defined below are intended to be mandatory to use or implement in all
+ * cases, but rather they provide a starting point for a set of useful,
+ * interoperable claims.
+ * Applications using JWTs should define which specific claims they use and when
+ * they are required or optional.
+ */
+export interface PayloadObject {
+  iss?: string;
+  sub?: string;
+  aud?: string[] | string;
+  exp?: number;
+  nbf?: number;
+  iat?: number;
+  jti?: string;
+  [key: string]: unknown;
+}
+
+export type Payload = PayloadObject | string;
+
+/*
+ * JWS §4.1.1: The "alg" value is a case-sensitive ASCII string containing a 
+ * StringOrURI value. This Header Parameter MUST be present and MUST be
+ * understood and processed by implementations.
+ */
+export interface Header {
+  alg: Algorithm;
+  [key: string]: unknown;
+}
+
+const encoder = new TextEncoder();
+const decoder = new TextDecoder();
+
+function isExpired(exp: number, leeway = 0): boolean {
+  return exp + leeway < Date.now() / 1000;
+}
+
+/*
+ * Helper function: setExpiration()
+ * returns the number of seconds since January 1, 1970, 00:00:00 UTC
+ */
+export function setExpiration(exp: number | Date): number {
+  return Math.round(
+    (exp instanceof Date ? exp.getTime() : Date.now() + exp * 1000) / 1000,
+  );
+}
+
+function tryToParsePayload(input: string) {
+  try {
+    return JSON.parse(input);
+  } catch {
+    return input;
+  }
+}
+
+export function decode(
+  jwt: string,
+): {
+  header: unknown;
+  payload: unknown;
+  signature: unknown;
+} {
+  const parsedArray = jwt
+    .split(".")
+    .map(base64url.decode)
+    .map((uint8Array, index) =>
+      index === 0
+        ? JSON.parse(decoder.decode(uint8Array))
+        : index === 1
+        ? tryToParsePayload(decoder.decode(uint8Array))
+        : convertUint8ArrayToHex(uint8Array)
+    );
+  if (parsedArray.length !== 3) throw TypeError("serialization is invalid");
+
+  return {
+    header: parsedArray[0],
+    payload: parsedArray[1],
+    signature: parsedArray[2],
+  };
+}
+
+export type TokenObject = {
+  header: Header;
+  payload: PayloadObject | string;
+  signature: string;
+};
+
+export function isTokenObject(object: TokenObject): object is TokenObject {
+  return (
+    typeof object?.signature === "string" &&
+    typeof object?.header?.alg === "string" &&
+    (typeof object?.payload === "object" && object?.payload?.exp
+      ? typeof object.payload.exp === "number"
+      : true)
+  );
+}
+
+export async function verify(
+  jwt: string,
+  key: string,
+  {
+    algorithm = "HS512",
+  }: {
+    algorithm?: Algorithm | Array<Exclude<Algorithm, "none">>;
+  } = {},
+): Promise<Payload> {
+  const obj = decode(jwt) as TokenObject;
```

Comment:
```suggestion
  const obj = decode(jwt);
```

Your guard at 119 narrows it to a `TokenObject`, coercing it here is unsafe.